### DISCUSSION TITLE
Locked iModel that cannot be unlocked.

### DISCUSSION THREAD

**Discussion by PeterJBassett** [iTwin/community#20] [Q&A]
Hello I am working on the FutureOn iTwin connector used by Shell. There is a problem that one of Shell's iModels is locked, and the usual methods of unlocking are not working.

The iModel is:
ContextId: c9e70df2-1133-44ae-9018-e2d4bb682fc6
iModel Id: 833f4130-22fa-459c-ac80-e9a8ea7863de

Attempting to sync to this iModel raises the following error log:
```
Error | itwin-connector.Framework | One or more objects have been locked in a newer Changeset. Object ids: 0x1
```

The connector is using the following iTwin versions:
```
  "dependencies": {
    "@itwin/connector-framework": "^2.0.5",
    "@itwin/core-backend": "^4.2.4",
    "@itwin/core-bentley": "^4.2.4",
    "@itwin/core-common": "^4.2.4",
    "@itwin/core-geometry": "^4.2.4",
    "@itwin/ecschema-metadata": "^4.2.4",
    "@itwin/imodels-access-backend": "^4.1.0",
    "@itwin/imodels-client-authoring": "^4.1.8",
```

**Comment by pmconne** [comment]
@jffmarker @swwilson-bsi @abeesh

**Reply by PeterJBassett** [reply]
Here is some more information about the problem.
It turned out that the synchronisation left some bim and bim-lock files in the local cache, and this is what was causing subsequent syncs to fail with the error shown above.

There shouldn't have been any files left in the local cache because my connector code always calls the following function after a either a successful sync or on any exception.

```
  /** cleanupCache: delete the iModel cached briefcase and changeset files */
  async cleanupCache() {
    try {
      // Get the iModel Briefcase details
      const briefcases = BriefcaseManager.getCachedBriefcases(this.config.iModelId)

      // Cleanup the cached briefcase files
      for (const briefcase of briefcases) {
        await BriefcaseManager.deleteBriefcaseFiles(briefcase.fileName, this.config.iModelAccessToken)
      }

      // Cleanup the cached changeset files
      await BriefcaseManager.deleteChangeSetsFromLocalDisk(this.config.iModelId)

    } catch(error) {
      console.log('Error deleting cached briefcase files')
    }
  }
```

So it seems that the BriefcaseManager.deleteBriefcaseFiles doesn't aleays work?

**Reply by akshay-madhukar** [reply]
That is good information to know. Thanks for relaying this back to us. I will take a look at BriefcaseManager and see why it is not deleting those bim files

**Reply by PeterJBassett** [reply]
Hi @akshay-madhukar A change I made recently was to add the optional access token parameter to `deleteBriefcaseFiles`. I thought this would be a benefit in unlocking the remote iModel at Bentley, but perhaps that is causing a failure sending the request, and then failing to remove the local cache briefcase files?

**Reply by akshay-madhukar** [reply]
Do you still have the bim-lock file that was causing the issue?

**Comment by PeterJBassett** [comment]
Hello Akshay

I’m sorry, but deleting the cached bim-lock file cleared the lock, and I didn’t think to keep a copy.

Peter

From: Akshay Madhukar ***@***.***>
Date: Tuesday, 6 February 2024 at 19:52
To: iTwin/community ***@***.***>
Cc: Peter Bassett ***@***.***>, Author ***@***.***>
Subject: Re: [iTwin/community] Locked iModel that cannot be unlocked. (Discussion #20)
CAUTION: EXTERNAL EMAIL. Don't click on links or open attachments unless you recognize the sender and know the content is safe.


Do you still have the bim-lock file that was causing the issue?
